[Concurrency] Add Swift Task Stealers

Bryce-MW opened the pull request:
When Swift Tasks are enqueued onto the cooperative pool, the queue they are placed onto is based on their current priority. If the Task is further escalated, it will be marked as such so that when it runs, it will self-escalate the thread to the required priority, but while it is enqueued, nothing will happen. If the Task was originally enqueued at a low priority, it may not be able to run because it is waiting on other items ahead and the system doesn’t have enough room to do low priority work. This is a priority inversion since this now-higher-priority Task is unable to run.
    
This commit implements a "stealer" object for Tasks that is similar to the existing Actor stealer called AsyncTaskStealer. These would be used only when Dispatch is the default global executor and priority escalation is enabled. To an executor of any kind (Dispatch or custom), stealer objects would look like any other Job. This avoids edge cases when a Task is enqueued to multiple types of executors through its lifetime (e.g. global, Actor, custom). Stealer objects have to be created any time the Task needs to be enqueued while the Task object itself is still enqueued somewhere (which in Dispatch executors is via an intrusive linked list thus only allowing the Task to be enqueued in one place). If it does eventually become dequeued, it may be enqueued directly the next time it needs to be enqueued. In order to ensure that the Task is only run when it should, it needs to contain a counter that increments each time it finishes an execution. Each stealer object contains a field indicating at which value of the counter is it able to run and the Task itself has one in its private data for the direct enqueued copy. If the Task or stealer is run but the Task’s current counter is not equal to the one that is allowed to be run, the invocation will do nothing.
    
Because flagAsRunning can now fail and a previous commit introduced an opaque return value, there is a split between two cases where flagAsRunning can be called. When the Task is being resumed immediately after suspension without being enqueued again, flagAsRunning is infailable and will always return zero for the opaque value. When initially running from swift_task_run, it is failable and may return an opaque value. I've refactored flagAsRunning to separate these cases and try my best to reduce some redundant code.
    
rdar://160967177

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Bryce-MW commented:
All three failures appear to be due to timeouts that don't appear related to my change but I'll do some more investigation

Bryce-MW commented:
@swift-ci smoke test

Bryce-MW commented:
@swift-ci smoke test

Bryce-MW commented:
@swift-ci smoke test

Bryce-MW commented:
@swift-ci smoke test

Bryce-MW commented:
@swift-ci smoke test

Bryce-MW commented:
@swift-ci smoke test

rokhinip commented on include/swift/ABI/Task.h:
```diff
@@ -430,21 +431,53 @@ class AsyncTask : public Job {
-  /// When Dispatch is used for the default executor:
-  /// * If the return value is non-zero, it must be passed
-  ///   to swift_dispatch_thread_reset_override_self
+  /// This function returns two values. The first, boolean, value
```
To make this comment easier to grok, it might be helpful to phrase it at a higher level like so, remove some of the heavy implementation details and redirect people to other places (including the implementation) for those details. Example alternative:

```
A task can become running either by first being enqueued, or directly from being suspended. 
With this change to task stealers, we mark this differentiation with `flagAsRunningFromSuspended` 
and `flagAsRunningFromEnqueued`

flagAsRunningFromEnqueued can be invoked from 2 contexts - from the intrusively linked task 
itself or from task stealers. As such, it needs to be failable such that only the Job with the 
correct allowedExclusionValue invokes the task successfully. If we failed to transition the 
task to running, the remaining setup needs to short-circuit and the task cannot be
 invoked by the caller.

flagAsRunningFromEnqueued's invokeFlags argument lets us know which context 
we are being called from which informs further decisions about lifetime 
management and future enqueue semantics of the task.

```

Bryce-MW committed:
[Concurrency] Fixups for qos_class_self() matching

In testing, a couple minor issues were discovered. Firstly, the property that we are trying to fix is that when running a Task, we let the operating system know what the
Task's basePriority is so that it may set the thread's base priority to match. The current base priority can be obtained via qos_class_self on Darwin platforms. However,
the code I had written had assumed that we could not know the current base priority and used the override floor as a guess of "the current thread's base priority must be
no greater than this". This was confusing and overcomplicated. Querying qos_class_self directly is fast and will give us exactly the correct value.

Secondly, the return value of dispatch_thread_override_self_with_base is going to be an opaque representation which includes the base priority of the thread or zero. We
want to use this value to reset the thread to what it originally was the first time that we called that function for this Task. However, if the CAS in flagAsRunning
fails, we may call this function again and it will return a value that is not what the thread originally was since we have already called it once before to change it.
If we have already received a value from the operating system, we should not overwrite it with another value because we would then be resetting the thread to an
intermediate state rather than the original state.

This also updates a few things in the dispatch shims file. One function had been mistakingly marked as available in a particular OS when it is not yet available in any
OS. This file also fails to build with an SDK which contains the required dispatch header but for which those functions do not yet exist. A macro is added to avoid
compiling in calls to those functions unless the dispatch header has defined the macro to a large enough value to indicate that they have been added to the header.

Bryce-MW commented:
@swift-ci smoke test

rokhinip commented on include/swift/ABI/Task.h:
```diff
@@ -795,12 +828,10 @@ class AsyncTask : public Job {
 private:
-  /// Access the next waiting task, which establishes a singly linked list of
-  /// tasks that are waiting on a future.
-  AsyncTask *&getNextWaitingTask() {
```
I am concerned about changing this - see more detailed comment in TaskPrivate.h

rokhinip commented on include/swift/ABI/TaskStatus.h:
```diff
@@ -409,35 +409,33 @@ class TaskDependencyStatusRecord : public TaskStatusRecord {
     EnqueuedOnExecutor,
   } DependencyKind;
 
-  // The task that has this task status record - ie a backpointer from the
```
See other comment in TaskPrivate for the 2 linkages in question

rokhinip commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -211,26 +212,26 @@ ExecutorTrackingInfo::ActiveInfoInThread;
+/// priority escalation is not enabled, this will always succeed.
+SWIFT_ALWAYS_INLINE
+static inline
+void taskInvokeWithExclusionValue(AsyncTask *task,
```
Yeah I'm increasingly convinced that removeEnqueued should become a set of invokeFlags conveying to this function the context in which it's being called.

rokhinip reviewed (commented):
I haven't finished reviewing the full PR yet but here are some preliminary comments and questions.

Bryce-MW commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1223,13 +1576,23 @@ inline void AsyncTask::destroyTaskDependency(TaskDependencyStatusRecord *depende
+  auto status = _private()._status().load(std::memory_order_relaxed);
+  assert(status.hasTaskDependency());
+#endif
+  auto *dependency = _private().dependencyRecord;
```
I'm happy to explain further but the crux of the problem is that with stealers, the Task can be both intrusively linked and suspended on another Task. So SchedulerPrivate can no longer ever be used by the runtime because at any state the Task may be intrusively linked because it was stolen.

I managed to figure out a way to deal with this without increasing the size of anything by sticking the await queue linkage into the dependency record and removing the backwards ref to the Task that wasn't really needed.

Specifically,
Previously, the dependency record would have the link to task T2 so that T1 may escalate it and a backwards reference to T1 itself so you could get from the record to the owning Task. The link for the linked list of Tasks waiting on T2 would be in SchedulerPrivate.
What I was able to do is remove the backwards reference to the owning Task because it was easy to work around needing that. I could then use that space to place the link for the Tasks waiting on T2. This frees SchedulerPrivate so that it is exclusively owned by the scheduler.

rokhinip commented on include/swift/Runtime/EnvironmentVariables.h:
```diff
@@ -58,6 +58,11 @@ using string = const char *;
 // Concurrency library can call.
 SWIFT_RUNTIME_STDLIB_SPI bool concurrencyEnableCooperativeQueues();
 
+// Wrapper around SWIFT_DEBUG_CONCURRENCY_ALWAYS_USE_STEALERS that the
```
Does this encompass both task and actor stealers? Should we differentiate?

Bryce-MW replied:
I decided to remove this for now because it is actually not useful as-is and I didn't test to know that it even works in this mode.

rokhinip commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -242,6 +243,48 @@ void swift::runJobInEstablishedExecutorContext(Job *job) {
     if (oldTask) ActiveTask::set(oldTask);
   } else {
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+    // This balances with the retain in taskEnqueueDirectOrSteal which happens
```
Let's chat offline about the lifetime changes here, it's a bit hard to grok from code

rokhinip commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -2748,8 +2788,22 @@ swift_actor_escalate(DefaultActorImpl *actor, AsyncTask *task, JobPriority newPr
+    SWIFT_TASK_DEBUG_LOG("Enqueuing stealer for %p on %p",
+                         (void*)task, (void*)executor.getIdentity());
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+    // See the comment in enqueueDirectOrSteal
```
Oh this is important. I would say that this should not be hidden in a comment here and should instead be part of a larger blurb added to the repo explaining where we have this feature vs not.

rokhinip commented on stdlib/public/Concurrency/TaskStatus.cpp:
```diff
@@ -312,6 +312,90 @@ void swift::removeStatusRecord(AsyncTask *task, TaskStatusRecord *record,
+
+  bool didRemove = false;
+  while (true) {
+    // If the record is locked, then either we wait for the lock or we own the
```
Nit: Feels more natural to tackle fast path (without needing a status record lock) first before the slowpath if we could reorganize the code to do so

rokhinip commented on stdlib/public/Concurrency/TaskStatus.cpp:
```diff
@@ -312,6 +312,90 @@ void swift::removeStatusRecord(AsyncTask *task, TaskStatusRecord *record,
+            if (didRemove) {
+              // If this record is the innermost record,
+              // set a new status with the record removed.
+              if (newStatus.getInnermostRecord() == record)
```
Once again, because the innermost record should be the status record lock, how could the record we're trying to remove be an innermost one?

Bryce-MW commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -242,6 +243,48 @@ void swift::runJobInEstablishedExecutorContext(Job *job) {
+
+  swift_release(stealer->Task);
+
+  swift_cxx_deleteObject(stealer);
```
This is how other non-Task Jobs work like ProcessOutOfLineJob, NullaryContinuationJob, and IsolatedDeinit work. I assume that it is refcounted but we never make more than one reference so it is ok to delete directly.

rokhinip commented on include/swift/ABI/Task.h:
```diff
@@ -182,6 +179,10 @@ class alignas(2 * alignof(void*)) Job :
     return RunJob(this); // 'return' forces tail call
   }
+
+  /// Gets the 32-bit Job ID from the job or the 64-bit
```
Yeap I found it later in the review. Thanks!

rokhinip commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -1494,6 +1537,9 @@ TaskExecutorRef TaskExecutorRef::fromTaskExecutorPreference(Job *job) {
   if (auto task = dyn_cast<AsyncTask>(job)) {
     return task->getPreferredTaskExecutor();
   }
+  if (auto stealer = dyn_cast<AsyncTaskStealer>(job)) {
```
So just to make sure I understand the scenario:

The task is intrusively linked somewhere but it proceeds to execute further due to a stealer running it. When it is running, it then asks to run something with an executor preference and then in that block inside `withTaskExecutorPreference` we escalate it?

rokhinip commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -992,97 +1038,395 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
+/// Remove the enqueued bit in the ActiveTaskStatus atomically.
+/// This must be done when a Task's intrusive link is dequeued
+/// but after reading the local stealer exclusion value.
+/// This should not be done from any concurrent context.
```
Well it can only be called by the task on itself right?

Bryce-MW commented on stdlib/public/Concurrency/TaskStatus.cpp:
```diff
@@ -1040,21 +1125,21 @@ static swift_task_escalateImpl(AsyncTask *task, JobPriority newPriority) {
     //
-    // TODO (rokhinip): Add a signpost to flag that this is a potential
-    //  priority inversion
+    // Still mark the task has having been escalated so that the
```
Putting a pin in this to take more of a look later because I need to remind myself why I thought I was correct

Bryce-MW commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -894,8 +938,10 @@ struct AsyncTask::PrivateStorage {
-      newStatus = newStatus.withoutEnqueued();
       newStatus = newStatus.withComplete();
+#if !SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+      newStatus = newStatus.withoutEnqueued();
```
(Another good reason why I should've split up the commits because then I could've written down why I did this)

rokhinip commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -992,97 +1038,390 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
+    if (!oldStatus.isEnqueued()) {
+      break;
+    }
+    // This could be an atomic AND rather than a CAS loop
```
Yeah we should do it with an atomic AND instead of a CAS

rokhinip commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -992,97 +1038,390 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
+};
+#endif
+
+static inline uint32_t taskFlagAsRunningWithoutDependency(AsyncTask &task, bool removeEnqueued = false) {
```
Is there a reason `taskFlagAsRunningWithoutDependency` is not a method on the task and is a free floating functions that takes a task?

rokhinip commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -992,97 +1038,390 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
-  return dispatchOpaquePriority;
+};
+
+/// Enqueue this task on the provided executor either directly or with a
```
I haven't had a chance to fully grok this function which feels loadbearing, I'll come back to it tomorrow.

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

mikeash commented on include/swift/ABI/Task.h:
```diff
@@ -94,7 +91,9 @@ class alignas(2 * alignof(void*)) Job :
+  // Reserved for the use of the scheduler. Since Task stealers
+  // allow a Task to be intrusively enqueued at any point of its
+  // lifecycle, this data must never be used by the runtime.
   void *SchedulerPrivate[2];
```
I'd like to see the do-not-touch-ness of this enforced a bit. At least a `DoNotUse` on the name, maybe even a custom type with private fields to make it really hard to inadvertently touch it.

Bryce-MW replied:
I'm not sure exactly what the best solution is here. There are a few legitimate uses like the CooperativeGlobalExecutor, Default Actor, a place where this is cleared, and to get the executor that dispatch is invoking from. I could see maybe making a custom type that has a method like `clearBeforeEnqueue` for the clear case and something like `getForExecutorUse<typename T>`? The type passed in could be a new type for each use that has more clear fields rather than the weird indexing system that exists now.

I think changing the name makes sense regardless because really any executor can use this, not just the "scheduler". I'm not totally clear on what scheduler means here but I think for example, for a Default Actor that Dispatch is still the scheduler but the Default Actor is allowed to use this field on the Tasks on its queue. It's the Actor's drainer Jobs that have this field reserved for Dispatch

Bryce-MW committed:
[Concurrency] Introduce Task Stealers

Adds an AsyncTaskStealer object and the machinery to enqueue a Task with a
stealer whenever intrusive enqueue isn't available, enabling stealers on the
regular enqueue path and on the executor escalation path.

Key changes:

 - AsyncTaskStealer: a new object that stands in for a Task in executor
   queues when the Task is already linked intrusively. Executor and task-ID
   lookups now look through the stealer.
 - Stealer exclusion value: a value stored on the Task that lets a stealer
   determine whether it still owns the right to run the Task; if the value
   has changed, the stealer bails out rather than double-running.
 - removeStatusRecordIf: a fallible variant of removeStatusRecord that
   refuses to remove the record when a condition has changed. Used as the
   syncronization point for the exclusion value.
 - IsEnqueued flag: replaced with a bit that records whether the Task is
   currently intrusively linked.
 - SchedulerPrivate is no longer used by the runtime while the Task is
   suspended. The dependency record's backlink is dropped and the current
   task is plumbed through the escalation path which is the only place the
   backlink was needed.
 - flagAsRunning refactor: splits out flagAsRunningFromSuspended (no
   exclusion value, no opaque return) and taskInvokeWithExclusionValue
   (which conditionally calls runInFullyEstablishedContext when
   flagAsRunningFromEnqueued succeeds). Both take invokeFlags so the
   stealer path is distinguishable. A new ThreadPriorityManager centralizes
   self-escalation and base-priority lowering, cutting duplication.
 - Adds a test that priority escalation works while a Task is enqueued,
   plus additional/corrected task debug logs.

Radar-Id: rdar://problem/160967177

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

rokhinip commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1110,6 +1110,100 @@ class AsyncTaskStealer : public Job {
+      auto newStealerExclusionValue = currentStealerExclusionValue;
+      if (__builtin_add_overflow(currentStealerExclusionValue, 1,
+                                 &newStealerExclusionValue) ||
+          newStealerExclusionValue > 0xF) {
```
Magic number -- let's replace this with a max value define or soemthing. 

I also think we should move this entire increment logic into into a new `withIncrementedStealerExclusionValue()` function inside ActiveTask Status and there, we can explain why we are capping it at 0xf cause of number of transitions of priority and the max number of bits we are using.

rokhinip commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1110,6 +1110,100 @@ class AsyncTaskStealer : public Job {
+    // This can always be relaxed because it only needs to be read
+    // either by a thread syncronizing with the status lock or someone
+    // reading the stealer which we will later publish on this thread.
+  } while ((updateStealerExclusionValue || !oldStatus.isIntrusivelyLinked()) &&
```
Why are `(updateStealerExclusionValue || !oldStatus.isIntrusivelyLinked())` part of the while loop condition?

rokhinip commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1110,6 +1110,100 @@ class AsyncTaskStealer : public Job {
+  if (needsStealer) {
+    // It is safe to read LocalStealerExclusionValue here becuase it can only
+    // be written to by this function which can not be called concurrently.
+    if (task->_private().LocalStealerExclusionValue ==
```
Ah okay so this is why you are actually needing that the task be status record locked? Urgh.

rokhinip commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -754,9 +750,20 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
+    // But to be more resilient, I get the offsets and types correctly
+    auto *flags = reinterpret_cast<std::atomic<decltype(ActiveTaskStatus::Flags)>*>(storage + offsetof(ActiveTaskStatus, Flags));
+    // Release memory ordering so that other threads can not decide to reuse this
+    // Task's intrusive linkage before all of our earlier uses have completed
```
Point of clarification: Are you trying to publish a write from a different memory location by the time this is published?

Bryce-MW replied:
I haven't strictly analyzed exactly what I care happens-before this. I am mainly imagining this as a sort of "lock" where we have "locked" the use of the Task intrusively so if another thread wishing to enqueue this Task sees that it is not enqueued intrusively after this line, that nothing before this line is doing to reorder after it in a disruptive way.

My general thinking is that since this is a release store and anyone who cares to enqueue the Task intrusively will be doing it under the Task Status Lock. Now interestingly enough, it will write to the locked bit of the Active Task Status using memory order consume which I don't really know the semantics of and if that will form the correct relationship with this release. I would be surprised if there isn't enough synchronization here but maybe it warrants closer analysis?

rokhinip commented on stdlib/public/Concurrency/TaskStatus.cpp:
```diff
@@ -312,6 +312,90 @@ void swift::removeStatusRecord(AsyncTask *task, TaskStatusRecord *record,
+      break;
+    }
+
+    // Nobody holds the lock, and the record is the
```
Similarly, this side feels like we should factor into a "fast path" function

rokhinip commented on include/swift/ABI/Task.h:
```diff
@@ -424,30 +424,51 @@ class AsyncTask : public Job {
-  /// the priority stored in the job flags if the priority has been
-  /// escalated.
+  enum InvokeFlags : uint32_t {
+    InvokedFromStealer = 0x01,
```
Do we not need a flag for InvokedFromTask?

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1110,6 +1110,100 @@ class AsyncTaskStealer : public Job {
 };
+
+/// Enqueue this task on the provided executor either directly or with a
+/// stealer. This must be called rather than swift_task_enqueue in case
```
I think that due to the fiddlyness of these functions, you probably won't be able to add a new callsite of this function or swift_task_enqueue without rewriting this function or at least changing the comment. So I'd rather keep it in TaskPrivate than put it in Concurrency.h because anyone who wants to do this operation again needs to be doing so through flagAsRunningFrom* unless they are us doing more careful work.

I've rewritten this comment in the next commit that I will post.

Bryce-MW commented on stdlib/public/Concurrency/TaskStatus.cpp:
```diff
@@ -312,6 +312,90 @@ void swift::removeStatusRecord(AsyncTask *task, TaskStatusRecord *record,
+    // lock. Either way, acquire the status record lock and perform the removal.
+    // If the record to be removed is not the innermost record, then we need to
+    // acquire the lock to safely remove it.
+    if (oldStatus.isStatusRecordLocked() ||
```
I think that refactoring status record manipulation functions should be a separate PR. removeStatusRecordIf is almost exactly copied from removeStatusRecord so if we change one, we should change both and I think modifying the existing one is out-of-scope for this PR

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

rjmccall commented on include/swift/ABI/Task.h:
I feel like this is restating what it means to be reserved for the scheduler. An executor is allowed to use these fields on a job between the points (1) when someone passes the job to the executor's `enqueue` and (2) when the executor either enqueues the job to another executor or passes it to `swift_job_run`. After point (2), the executor is no longer allowed to touch the fields, not because of any complicated argument about task stealers but just because the job's allowed to pass itself to `enqueue`.

Are we arguing that the runtime is not allowed to touch the fields even when it is acting as an executor? Like, a default actor cannot use these fields to managed the queue of jobs that are enqueued on an actor? I don't understand why that would be either necessary or a good trade-off for us to make; actor enqueuing is common.

Bryce-MW replied:
The semantics haven't changed from the perspective of a scheduler. The scheduler still has exclusive use of the field in the situations you mentioned. The default actor can for sure use this field on any Job enqueued on it

What has changed is that the amount of time an executor has exclusive use of this field has increased for Tasks specifically. Because a stealer could exist, this field on a Task could be used by the executor it was intrusively linked to even if the Task is suspended or goes through any other part of it's lifecycle

Previously, if a Task was suspended waiting on another Task, this field was reused to be the linked list of waiting Tasks. But now, this field might still be used by an executor even while the Task is suspended so the runtime can no longer user it for this linked-list purpose

So I suppose that it's not really never used by the runtime since the runtime can implement an executor (like the default actor), it just can't be used for non-executor purposes

Bryce-MW replied:
I'll update this comment to be more clear about what I mean

rjmccall replied:
Thanks, this is good.

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -407,6 +410,43 @@ class AsyncTask : public Job {
+    InvokeFlagsFromStealer = 0x01,
+  };
+  /// Flag that this task is now running. This can update the priority
+  /// stored in the job flags if the priority has been escalated.
```
I feel like the name (and this leading sentence of the comment) should reflect that this can now fail. And "flag" is starting to feeling pretty weak for all the things this does. Maybe `tryStartRunning`?

I think the comment needs to start by explaining the whole semantic operation, something like:

> Try to flag that the current thread is going to run this task. If there are multiple jobs trying to run the task, this requires claiming the right to run the task, which can fail if another job wins the race. This may also require increasing changing the priority of the current thread, if the current task executor supports that.

After that, you can break down the details:
- how failure is reported back to the caller, and what the caller's responsibilities are in that case
- the special management of the intrusively linked task
- the rules around the priority override token

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -407,6 +410,43 @@ class AsyncTask : public Job {
+  /// the thread before returning to the context invoking this Task (either the
+  /// executor or the caller of Task.immediate). When flagAsRunning succeeds,
+  /// Dispatch is the default executor, and priority escalation is enabled,
+  /// this value must be passed to swift_dispatch_thread_reset_override_self.
```
This sounds like a great place to use a typedef (or better yet, a `struct`) to help communicate and encapsulate the purpose of that value.

Bryce-MW replied:
I was thinking I could put something like struct swift_dispatch_priority_token in DispatchShims.h but most of that code only gets compiled in if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION is enabled. So I'd have to put the wrapper above the check or put it in Concurrency.h. Do you happen to have a suggestion on where best to put this? I like the idea though because that would make it a lot more clear

rjmccall replied:
You could define it above the check, I think that's reasonable.

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -407,6 +410,43 @@ class AsyncTask : public Job {
+  /// Dispatch is the default executor, and priority escalation is enabled,
+  /// this value must be passed to swift_dispatch_thread_reset_override_self.
+  std::pair<bool, uint32_t>
+  flagAsRunningFromEnqueued(uint8_t allowedExclusionValue,
```
Maybe exclusion values should also be wrapped up as structs?

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -407,6 +410,43 @@ class AsyncTask : public Job {
+  ///
+  /// A task can become running either by first being enqueued, or directly
+  /// from being suspended. There is a separate function for each of these
+  /// scenarios, `flagAsRunningFromSuspended` and `flagAsRunningFromEnqueued`.
```
I would say something more like:

> This function is for when the task starts running after fully suspending. If a task starts to suspend itself, but then finds out that it didn't need to, it should instead call `flagAsRunningFromSuspended`, which avoids a lot of the extra complexity of this operation.

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -421,30 +461,17 @@ class AsyncTask : public Job {
+  /// where you may mark yourself as suspended but find out during
+  /// atomic state update that you may actually resume immediately.
+  void
+  flagAsRunningFromSuspended(InvokeFlags invokeFlags = InvokeFlagsFromTask);
```
This is a much more specialized operation, and I would give it a longer name that better distinguishes it, something like `resumeRunningAfterFailedSuspend`. With the names as they are in this PR, I honestly wouldn't blink if I saw something calling the wrong operation, and that's really bad.

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -820,12 +849,10 @@ class AsyncTask : public Job {
-  }
+  /// Access the next waiting task, which establishes a singly linked
+  /// list of tasks that are waiting on a future. This function
+  /// assumes that this Task is suspended waiting on a another Task.
```
```suggestion
  /// assumes that this Task is suspended waiting on another Task.
```

rjmccall commented on include/swift/Runtime/DispatchShims.h:
```diff
@@ -51,9 +60,13 @@ swift_dispatch_thread_override_self(qos_class_t override_qos) {
     return dispatch_thread_override_self_with_base(override_qos, base_qos);
-  } else if (__builtin_available(macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0, *)) {
+  } else
+#endif
```
I see what you're doing, but it's pretty frightening to end a `#if` with a trailing `else` this way, especially an unbraced `else`. Fortunately, the block you're putting in the `#if` is actually an early exit, so you actually don't need this to be an `if/else` chain at all.

Bryce-MW replied:
Yea, I really hated that. Totally forgot that it doesn't need an else since it has the short-circuit. Very glad to fix that!

rjmccall commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -2724,22 +2758,19 @@ static void swift_task_enqueueImpl(Job *job, SerialExecutorRef serialExecutorRef
-    if (auto task = dyn_cast<AsyncTask>(job)) {
-      auto taskExecutorRef = task->getPreferredTaskExecutor();
-      if (taskExecutorRef.isDefined()) {
+    auto taskExecutorRef = TaskExecutorRef::fromTaskExecutorPreference(job);
```
Surely this is the same value as `_taskExecutorRef` above?

Bryce-MW replied:
I mostly was doing a find and manual replace job here. I think the idea what that getPreferredTaskExecutor would walk the status records so it wouldn't be done if it wasn't needed. The above calculation happens only in NDEBUG so in release builds, the walk would only happen for non-default serial executors. I'm happy to just push it up though so it only needs to be done

rjmccall replied:
Oh, somehow I overlooked that it was still in NDEBUG. This is fine, yeah, let's avoid doing the work if it's unneeded.

Bryce-MW commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1129,82 +1640,114 @@ AsyncTask::flagAsAndEnqueueOnExecutor(SerialExecutorRef newExecutor) {
+
+          // Set up task for enqueue to next location
+          // by setting the Job priority field
+          Flags.setPriority(lockedStatus.getStoredPriority());
```
The two priority updates here need to be moved into swift_task_enqueueSelfOrStealer. Because on escalate enqueue of the Task intrusive link, this won't happen so the Task's Job won't have the correct priority value. On regular enqueue of a stealer, this will also do an extra change to the Task's JobFlags which isn't a problem and we do set stealer priority correctly but is a bit weird.

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -457,17 +499,14 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
+    /// This bit refers to the enqueued-ness on the specific queue that
+    /// the task is linked into and therefore, can only be cleared when
+    /// the intrusively linkage is cleaned up. The enqueued-ness is
+    /// orthogonal to the other states of running/suspended/completed.
```
I think `IsIntrusivelyEnqueued` or `IsDirectlyEnqueued` might be better names. All of this discussion about intrusive linking seems over-indexed on a specific issue; it's really about whether there's some executor that might try to run this specific task as a job.

You also need to be explicit about what this does and does not permit, either here or somewhere more fundamental (like the main comment on either `Job` or `Task` in `Task.h`?), something like this:

> In general, the header fields of a job that's currently enqueued must not be mutated. The executor owns that memory and expects it to be immutable or (in the case of the schedule-private storage) under its control until it dequeues the job. Once the executor calls `swift_job_run`, the job regains ownership of the header fields and is generally permitted to change them again.
>
> Since `Task`s are often repeatedly enqueued as jobs, some extra care must be taken with them because of the runtime's use of task stealer jobs. When a task's priority is increased while it is enqueued, the runtime will enqueue a job with the higher priority on the same executor to try to take over the responsibility of running the task. This means that the `Task` object (considered as a `Job`) may still be enqueued on some executor even while the task is logically running, because the task may have started running on behalf of a task stealer job.  So while code that's acting on behalf of the current running task is free to assume exclusive access to most of the `Task` structure, it cannot generally modify the task's `Job` header unless it knows that the task object is not presently enqueued as a job this way. (This is tracked dynamically in the task status.)
>
> However, an important exception to that rule exists for the task resume function and context. Importantly, these fields are not directly accessed by executors. Tasks and task stealer jobs coordinate so that only one of them will advance the task at once. This coordination establishes a rule that any outstanding jobs (possibly including the `Task` itself) that were enqueued for previous suspensions of the task will never actually start running long enough to access these fields. As a result, code running on behalf of the current task can freely set the resume fields before suspending.

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -485,8 +524,40 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
+    // anyone other than the executor. To ensure it is not disposed until
+    // after it is finally dequeued, an additional retain is added when
+    // the first stealer is enqueued which this bit tracks. The balancing
+    // release happens when the intrusively linked Task is dequeued.
```
I don't think the scheme described here works if multiple stealers are created.

Bryce-MW replied:
Each TaskStealer holds an additional refcount to the Task. This is for the case where a Task is enqueued, a TaskStealer is enqueued, and through the stealer, the Task entirely completes and the client releases all of its references. Tasks don't get an additional refcount when enqueued directly (and it's not easy to add one). Before, a Task could never be completed and directly enqueued so this wasn't a problem because a refcount is held by a Task being incomplete even if no client holds a refcount to the Task. However, since a Task could now complete and have no enqueued stealers but remain directly enqueued, it would have zero refcount.

My solution is that if you ever enqueue a stealer, try to set this bit and if you do set it, take an extra refcount. This gets dropped when the Task tries to run. I also think of this as "add a retain if this Task's Job could fail to run" or "add a retain for the first additional stealer added after the last enqueue of this Task's Job"

There's probably a better comment that I could write to explain this but I couldn't think of it

rjmccall replied:
The way we handle a similar problem with default actors is that creating a stealer increments and losing the race to run the task decrements. So if a stealer wins the race, ownership of its refcount mystically transfers to the task.

Bryce-MW replied:
That's very similar to what I had originally implemented but Rokhini suggested that using an explicit bit would be more clear even though it's not strictly required. She did say that dispatch pulls a bunch of refcount tricks so the clever method wouldn't be unusual but needs to be documented well

rjmccall replied:
So this is still an extra retain/release, right? And yeah, we're only doing it for the first stealer we enqueue at once, but that's ~100% of the time going to be the only stealer we enqueue, so it's two extra atomic operations essentially every time this kicks in.

Since we already use the decrement-on-loss approach elsewhere, I feel we ought to just explain it well once, then use it in multiple places rather than using different schemes in different parts of the code base.

...shame that it doesn't completely remove the need to type-pun the atomic object, though.

Bryce-MW replied:
I think it is technically possible to implement without needing a bit in here at all (and thus not requiring the atomic AND). That was what I did in my initial implementation though mine did still have the extra retain-release.

What I'm thinking is that it could have the pattern you suggested where stealers retain and stealers that loose release. Thus a winning stealer transfers its retain to the Task. The Task could then decrement if it looses the race since that can only happen if a stealer wins. However, there is an extra complication. If the Task is stuck at low priority, multiple winning stealers could happen but the task would only eventually loose once. What could be done is that a stealer only decides not to decrement if it both wins _and_ has the same exclusion value as the Task itself.

On the other hand, getting this lifetime right was one of the most tricky parts of the development and required a lot of careful live-on to track down bugs. I'm a bit weary of making another change in this area for the initial implementation, especially one that makes the scheme more complicated. I'd prefer to resolve it in a follow-up so that the initial version can go in sooner and I can get the ~month of live-on that would make me feel comfortable with a change here

rjmccall replied:
I'm fine with doing this optimization as a follow-up, sure.

Bryce-MW replied:
Filed rdar://181852142 (Improve retain-release scheme for a Task with Stealers) to do it

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -485,8 +524,40 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+    /// The Task's intrusive link or a Stealer may only run if its
+    /// exclusion value is equal to this value. This number increases
+    /// only when running a Task after it was escalacted while it
```
```suggestion
    /// only when running a Task after it was escalated while it
```

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -485,8 +524,40 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
+    ///
+    /// There are only 5 values of JobPriority and this value can only
+    /// increase at most once for each priority transition so this
+    /// could just use 3 bits (along with PriorityMask) if needed.
```
Hmm. It seems reasonable to rely on this for correctness, but I think I'd be more comfortable if we enforced it somehow. I guess that would look something like only initiating the stealing process if priority hit the next bucket?

Bryce-MW replied:
Yea, the concern here is that I think there's no enforcement that JobPriority (on Darwin) is only one of the acceptable values. There are a bunch of places where we cast it to qos_class_t though so really making it robust to other values would require additional effort. For any Dispatch executor, if you try to enqueue a Task at a non-QoS JobPriority value, it will crash (in most cases).

I did do an overflow check when incrementing this value such that on assertions builds it will assert. In a release build, this would overflow from F back to 0. The default global executor has less than F acceptable values (other than the real QoS values, it will also accept 0 and 2 returning the default and user-initiated queues, we should clean up getGlobalQueue to not do this). Given that we only enqueue stealers on escalation when on the global, I think it's impossible to overflow in practise without crashing due to enqueuing a Task with a bad priority.

I think figuring out how to make this more robust is a good idea. Do you want me to try making some part of this more robust in this PR?

rjmccall replied:
Hmm. Since we are tied to the global, I'm fine with just relying on those checks for now. But yeah, in the long run it does feel like we should either not make assumptions about valid priorities or make sure we canonicalize priorities earlier.

NotTheNHK replied:
There were discussions about this on the forums a while back regarding whether we should canonicalize "bad" priorities for the different platforms executors, with a semi-related issue tracked here: https://github.com/swiftlang/swift/issues/88584.

Bryce-MW replied:
Besides the direct crash, the code to override a running thread (either the self-override now in ThreadPriorityManager::overrideIfNeeded or the indirect override in swift_task_escalateImpl) casts to qos_class_t. I don't think this crashes but it won't do anything reasonable. This makes me feel like this should lean towards being an open problem that needs to be thought through

NotTheNHK replied:
FYI, I’ve started working on this here: #90000.

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -485,8 +524,40 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
+    /// could just use 3 bits (along with PriorityMask) if needed.
+    StealerExclusionShift = 20,
+    StealerExclusionMask = 0xF00000,
+    StealerExclusionMax = 0xF,
```
`StealerExclusionMask = StealerExclusionMax << StealerExclusionShift`

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -536,75 +615,72 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
   }
-  bool isCancelledIgnoringShield() const { return Flags & IsCancelled; }
-  ActiveTaskStatus withCancelled() const {
-#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
```
Can you make sure that you don't reformat unrelated code or whatever is causing all these lines that you didn't really semantically change to show up as differences?

Bryce-MW replied:
With this part, I was adding a new getter-setter and needed to determine where to put it. I realized that there was a ton of duplication with all of the #ifs and there was no consistent ordering to them. So I made a withFlags abstraction and reordered them to all be in the order that they are in the statue. This was helpful for me while developing to find which implementation I was looking for. But it's not important so I could remove it

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -619,115 +695,137 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
+    // But to be more resilient, I get the offsets and types correctly
+    auto *flags =
+        reinterpret_cast<std::atomic<decltype(ActiveTaskStatus::Flags)> *>(
+            storage + offsetof(ActiveTaskStatus, Flags));
```
Why can this not be done normally, using standard access patterns for the task status atomic?

Bryce-MW replied:
I wanted to do an atomic AND here which is a single instruction rather than a CAS loop. However, `std::atomic` will only allow fetch_and if the type is an integer which ActiveTaskStatus isn't. This is the "correct" way to do this cast in C++ though practically the syntax could be simpler if I assume the type is always uint32_t and offset is always zero.

rjmccall replied:
Hmm, alright. I don't know that this operation is hot enough to be worth this, but I'll concede that we're not likely to get anything like this code pattern without taking it into our own hands.

rjmccall reviewed (commented):
A bunch of comments for now.

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1036,97 +1152,515 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
+swift_task_enqueueSelfOrStealer(AsyncTask *task, SerialExecutorRef newExecutor,
+                                EnqueueFlags flags) {
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+  // Assertions / Logging, these intitial values need not have
```
```suggestion
  // Assertions / Logging, these initial values need not have
```

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1036,97 +1152,515 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
+
+/// When this task has a valid dependent on executor record, and is
+/// holding the TaskStatusLock, this function will enqueue the Task on
+/// the provided executor either directly or with a steaaler if needed.
```
```suggestion
/// the provided executor either directly or with a stealer if needed.
```

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1036,97 +1152,515 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
+  EnqueueFlagsForEscalation = 0x1,
+};
+static inline void
+swift_task_enqueueSelfOrStealer(AsyncTask *task, SerialExecutorRef newExecutor,
```
When I read this patch, I keep double-taking when I see that calls to this are conditional, because enqueuing the task doesn't feel like something that we should ever conditionally do. Maybe it makes sense to have different versions of this (even if they ultimately use the same implementation) for mandatory enqueues vs. priority-escalation enqueues.

Also, same comment about this not belonging in a header.

Bryce-MW replied:
There are only three calls to this function and the only conditional one is in swift_executor_escalate. I'm not entirely sure what it would look like to have different versions but I'm happy to make it more clear. Maybe a different name or better comments somewhere?

I think I originally put it in this file because it was only used by functions in this file. But it does make more sense to put in Actor.cpp alongside swift_task_enqueueImpl

rjmccall replied:
Just adding a comment to that call from the escalation path is probably okay.

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1036,97 +1152,515 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
+  if (isAsyncLet) {
+    // Async let tasks allocate within their parent's allocator so they
+    // may not outlive their parent which is possible if a stealer is
+    // introduced. Thus, if we are on the escalation path, return immediatly
```
I think you're missing the really important part of this explanation; consider something more like:

> async let tasks are allocated on the async stack of their parent task,
> so the task objects can't outlive the scope of the let binding. Task
> stealers can allow us to finish the task while the task (or a superseded
> stealer) is still enqueued somewhere, so this is a problem. We don't
> want to delay waking up the parent task and exiting the async let
> scope until all these extra references resolve themselves, either.
> For now, just don't ever enqueue task stealers for async let tasks,
> even though this means we can still have priority inversions when
> they're enqueued.

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -1036,97 +1152,515 @@ inline bool AsyncTask::isCancelled(bool ignoreShield = false) const {
+}
+
+inline std::pair<bool, uint32_t>
+AsyncTask::tryStartRunning(AsyncTask::ExclusionValue allowedExclusionValue,
```
This operation clearly doesn't belong in a header. There isn't any imaginable world in which this is going to get inlined.

Bryce-MW replied:
This is where flagAsRunning was defined before. I'd generally agree that non-templated functions shouldn't be in headers. Do you want me to move it in this PR? I would be inclined to move the related functions at the same time but some like flagAsSuspended I didn't really modify so it would make for a bigger diff

rjmccall replied:
I'm just thinking about the size of the function. This was originally in a header because it wasn't doing anything besides touching the status, and it's gradually turned into a pretty large and eventful function. But you can do this in a follow-up if you like.

Bryce-MW replied:
Filed rdar://181852536 (Clean up functions from TaskPrivate.h that should be in Actor.cpp) to do this

rjmccall reviewed (commented):
Okay. Mostly style / naming / comment nit-picks. I would like us to use the release-on-loss refcounting scheme, since we're already using that elsewhere. If we don't have a good explanation of the scheme in the code base already, we already need to fix that, and then we can just cite it twice.

rjmccall commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -2799,15 +2837,29 @@ swift_actor_escalate(DefaultActorImpl *actor, AsyncTask *task, JobPriority newPr
 
+  // All dispatch custom executors and every custom executor
+  // implementation I've seen so far would support (or at least
+  // not break) with adding stealers with the current semantics.
```
I think what you're trying to say is something more like:

```
// If the task is enqueued on a default actor, escalate the actor.
...
// If the task is enqueued on any other executor, we can enqueue
// a task stealer job on that executor. For now, don't do this if it's
// a custom executor, just in case the custom executor has a
// problem with enqueuing non-task jobs.
```

With that said, I don't think we really need or want to be conservative about this except as an implementation staging thing. The logic works for arbitrary priority-honoring executors, and they really are supposed to handle arbitrary jobs and not just tasks.

We may want to offer some kind of official API to `ExecutorJob` for querying if the job is a task stealer so that custom executors can recognize stealers and handle them specially. That can certainly be future work, though.

Bryce-MW replied:
Filed rdar://181848842 (Allow Task Stealers for escalation on custom executors) to remove the limitation

I did test without it and didn't encounter any issues so I'll be happy to remove the check

rjmccall commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -627,115 +703,138 @@ class alignas(2 * sizeof(void*)) ActiveTaskStatus {
 
-  JobPriority getStoredPriority() const {
-    return JobPriority(Flags & PriorityMask);
+  // HasActiveTaskCancellationShield
```
This comment seems useless.

rjmccall reviewed (approved):
Alright, I'm okay with this PR as it stands, with some nits about comments. Please do move the large functions from the header at some point.

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test

Bryce-MW commented:
@swift-ci please smoke test Windows

Bryce-MW commented:
@swift-ci please smoke test Windows
